internal/sfio/migrate.go

```diff
@@ -0,0 +1,212 @@
+		yaml.Tee(
+			yaml.Lookup("spec"), ClearFieldComment("template", "# trial"),
+			yaml.Lookup("spec"), ClearFieldComment("template", "# job"),
+			yaml.Lookup("spec"), ClearFieldComment("template", "# pod"),
+		),
+
+		// Rename the template fields
+		yaml.Tee(
+			yaml.Lookup("spec"), RenameField("template", "trialTemplate"),
+			yaml.Lookup("spec"), RenameField("template", "jobTemplate"),
+		),
+
+		// Migrate the parameters
+		yaml.Tee(
+			yaml.Lookup("spec", "parameters"), yaml.FilterFunc(f.migrateParametersV1alpha1),
+		),
+
+		// Migrate the metrics
+		yaml.Tee(
+			yaml.Lookup("spec", "metrics"), yaml.FilterFunc(f.migrateMetricsV1alpha1),
+		),
+
+		// Finally, set the apiVersion
+		yaml.Tee(
+			yaml.SetField("apiVersion", yaml.NewStringRNode(redskyv1beta1.GroupVersion.String())),
+		),
+	)
+}
+
+func (f *ExperimentMigrationFilter) migrateParametersV1alpha1(node *yaml.RNode) (*yaml.RNode, error) {
+	elements, err := node.Elements()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, e := range elements {
+		err := e.PipeE(
+			// If values is non-empty, remove min/max
+			yaml.Tee(
+				Has(yaml.Lookup("values", "-")),
+				yaml.Tee(yaml.Clear("min")),
+				yaml.Tee(yaml.Clear("max")),
+			),
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return node, nil
+}
+
+func (f *ExperimentMigrationFilter) migrateMetricsV1alpha1(node *yaml.RNode) (*yaml.RNode, error) {
+	elements, err := node.Elements()
+	if err != nil {
+		return nil, err
+	}
+
+	for _, e := range elements {
+		m := metric{}
+		if err := e.YNode().Decode(&m); err != nil {
+			return nil, err
+		}
+
+		err := e.PipeE(
+			// Change metric type "local" to "kubernetes"
+			yaml.Tee(
+				yaml.MatchField("type", "local"),
+				yaml.Set(yaml.NewStringRNode(string(redskyv1beta1.MetricKubernetes))),
+			),
+
+			// Change type "pods" to "" and add "target: { kind: PodList }"
+			yaml.Tee(
+				Has(
+					yaml.MatchField("type", "pods"),
+					yaml.Set(yaml.NewStringRNode("")),
+				),
+				yaml.SetField("target", yaml.NewMapRNode(&map[string]string{
+					"kind": "PodList",
+				})),
+			),
+
+			// Set or clear the URL field
+			yaml.Tee(
+				m.setURLField("url"),
+			),
+
+			// Rename the selector to target (must happen AFTER setting the "url" field because we unmarshal it)
+			yaml.Tee(
+				RenameField("selector", "target"),
+			),
+		)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return node, nil
+}
+
+// metric represents the parts of a v1alpha1 metric required for producing a v1beta1 metric URL.
+type metric struct {
+	Type     string                `yaml:"type"`
+	Scheme   string                `yaml:"scheme"`
+	Selector *metav1.LabelSelector `yaml:"selector"`
+	Port     intstr.IntOrString    `yaml:"port"`
+	Path     string                `yaml:"path"`
+}
+
+// setURLField returns a filter that will set the named field with the metric URL.
+func (m *metric) setURLField(name string) yaml.Filter {
```

Comment:
is there any of this that we can reuse/migrate/refactor from https://github.com/thestormforge/optimize-controller/blob/master/api/v1alpha1/experiment_conversion.go#L105

Reply:
I mentioned it in the description, that code is going away (along with the rest of the `v1alpha1` package). Once this PR is in, I'll cut a v2 branch and start poking at those changes (I want this fix command in both 1.x and 2.x though).